File: sdks/node/src/plugins/droid-io.ts
import os from 'node:os';
import path from 'node:path';
import { MAX_SOURCE_BYTES } from './common.js';
import { AGENT_KEY, type RuntimeContract } from './droid-contract.js';
import {
  activeDocument,
  createLegacyHookDocument,
  createOwnedHookDocument,
  createSettingsDocument,
  hookBlock,
  parseDocument,
  type DroidDocument,
  type DroidDocumentKind,
  type DroidSources,
} from './droid-config.js';
import { readDroidPolicy } from './droid-policy.js';
import { inspectPhysicalDirectory, readPhysicalFile } from './managed-files.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';

interface FactoryPaths {
  readonly directory: string;
  readonly root: string;
  readonly legacyDirectory: string;
  readonly legacy: string;
  readonly settings: string;
  readonly localSettings: string;
}

function factoryPaths(): FactoryPaths {
  const directory = path.join(os.homedir(), '.factory');
  const legacyDirectory = path.join(directory, 'hooks');
  return {
    directory,
    root: path.join(directory, 'hooks.json'),
    legacyDirectory,
    legacy: path.join(legacyDirectory, 'hooks.json'),
    settings: path.join(directory, 'settings.json'),
    localSettings: path.join(directory, 'settings.local.json'),
  };
}

async function readDocument(
  filePath: string,
  kind: DroidDocumentKind,
  label: string,
): Promise<DroidDocument | undefined> {
  const snapshot = await readPhysicalFile(filePath, label, MAX_SOURCE_BYTES);
  return snapshot ? parseDocument({
    exists: true,
    filePath,
    kind,
    raw: snapshot.contents,
    snapshot,
  }) : undefined;
}

export async function readSources(): Promise<DroidSources> {
  const paths = factoryPaths();
  await inspectPhysicalDirectory(paths.directory, 'Factory Droid user configuration directory');
  await inspectPhysicalDirectory(paths.legacyDirectory, 'Factory Droid legacy hooks directory');
  const [root, legacy, settings, localSettings, policy] = await Promise.all([
    readDocument(paths.root, 'hooks', 'Factory Droid user hooks'),
    readDocument(paths.legacy, 'legacy', 'Factory Droid legacy hooks'),
    readDocument(paths.settings, 'settings', 'Factory Droid user settings'),
    readDocument(paths.localSettings, 'local-settings', 'Factory Droid local settings'),
    readDroidPolicy(),
  ]);
  return {
    root: root ?? createOwnedHookDocument(paths.root),
    legacy: legacy ?? createLegacyHookDocument(paths.legacy),
    settings: settings ?? createSettingsDocument(paths.settings),
    localSettings: localSettings
      ?? createSettingsDocument(paths.localSettings, 'local-settings'),
    policy,
  };
}

export function requireHooksEnabled(sources: DroidSources): void {
  const blocked = hookBlock(sources);
  if (blocked) {
    throw new Error(
      `Factory Droid user hooks are disabled by ${blocked.field} in ${blocked.label} at ${blocked.filePath}`,
    );
  }
}

export async function runtimeFilesExist(contracts: RuntimeContract[]): Promise<boolean> {
  for (const contract of contracts) {
    if (await managedRuntimeFilesExist(contract, AGENT_KEY)) return true;
  }
  return false;
}

export function displayConfigPath(sources: DroidSources): string {
  return activeDocument(sources).filePath;
}

File: sdks/go/cmd/elydora/plugins/droid_io.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
)

type droidFactoryConfigPaths struct {
	directory       string
	root            string
	legacyDirectory string
	legacy          string
	settings        string
	localSettings   string
}

func droidFactoryPaths() (*droidFactoryConfigPaths, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("resolve home directory: %w", err)
	}
	directory := filepath.Join(home, ".factory")
	legacyDirectory := filepath.Join(directory, "hooks")
	return &droidFactoryConfigPaths{
		directory:       directory,
		root:            filepath.Join(directory, "hooks.json"),
		legacyDirectory: legacyDirectory,
		legacy:          filepath.Join(legacyDirectory, "hooks.json"),
		settings:        filepath.Join(directory, "settings.json"),
		localSettings:   filepath.Join(directory, "settings.local.json"),
	}, nil
}

func readDroidDocument(
	filePath, kind, label string,
) (*droidDocument, error) {
	snapshot, err := readManagedFile(filePath, label, maxManagedSourceBytes)
	if err != nil || snapshot == nil {
		return nil, err
	}
	return parseDroidDocument(
		true,
		filePath,
		kind,
		snapshot.contents,
		snapshot,
	)
}

func readDroidSources() (*droidSources, error) {
	paths, err := droidFactoryPaths()
	if err != nil {
		return nil, err
	}
	if _, err := managedPhysicalDirectoryExists(
		paths.directory,
		"Factory Droid user configuration directory",
	); err != nil {
		return nil, err
	}
	if _, err := managedPhysicalDirectoryExists(
		paths.legacyDirectory,
		"Factory Droid legacy hooks directory",
	); err != nil {
		return nil, err
	}
	root, err := readDroidDocument(paths.root, "hooks", "Factory Droid user hooks")
	if err != nil {
		return nil, err
	}
	if root == nil {
		root, err = createOwnedDroidDocument(paths.root)
		if err != nil {
			return nil, err
		}
	}
	legacy, err := readDroidDocument(paths.legacy, "legacy", "Factory Droid legacy hooks")
	if err != nil {
		return nil, err
	}
	if legacy == nil {
		legacy, err = createDroidDocument(paths.legacy, "legacy", []byte("{}\n"))
		if err != nil {
			return nil, err
		}
	}
	settings, err := readDroidDocument(paths.settings, "settings", "Factory Droid user settings")
	if err != nil {
		return nil, err
	}
	if settings == nil {
		settings, err = createDroidDocument(paths.settings, "settings", []byte("{}\n"))
		if err != nil {
			return nil, err
		}
	}
	localSettings, err := readDroidDocument(
		paths.localSettings,
		"local-settings",
		"Factory Droid local settings",
	)
	if err != nil {
		return nil, err
	}
	if localSettings == nil {
		localSettings, err = createDroidDocument(
			paths.localSettings,
			"local-settings",
			[]byte("{}\n"),
		)
		if err != nil {
			return nil, err
		}
	}
	policy, err := readDroidPolicy()
	if err != nil {
		return nil, err
	}
	return &droidSources{
		root:          root,
		legacy:        legacy,
		settings:      settings,
		localSettings: localSettings,
		policy:        policy,
	}, nil
}

func requireDroidHooksEnabled(sources *droidSources) error {
	if blocked := droidHookBlocked(sources); blocked != nil {
		return fmt.Errorf(
			"Factory Droid user hooks are disabled by %s in %s at %s",
			blocked.field,
			blocked.label,
			blocked.filePath,
		)
	}
	return nil
}
